velox/dwio/dwrf/test/WriterTests.cpp

```diff
@@ -79,8 +81,114 @@ class WriterTest : public Test {
+ public:
+  AllWriterCompressionTest() : compressionKind_(GetParam()) {}
+
+  static std::vector<CompressionKind> getTestParams() {
+    static std::vector<CompressionKind> params = {
+        CompressionKind::CompressionKind_NONE,
+        CompressionKind::CompressionKind_ZLIB,
+        CompressionKind::CompressionKind_SNAPPY,
+        CompressionKind::CompressionKind_LZO,
+        CompressionKind::CompressionKind_ZSTD,
+        CompressionKind::CompressionKind_LZ4,
+        CompressionKind::CompressionKind_GZIP,
+        CompressionKind::CompressionKind_MAX};
+    return params;
+  }
+
+ protected:
+  CompressionKind compressionKind_;
+};
+
+TEST_P(AllWriterCompressionTest, compression) {
+  std::map<std::string, std::string> overrideConfigs;
+  overrideConfigs.emplace(
+      Config::COMPRESSION.configKey(), std::to_string(compressionKind_));
+  auto config = Config::fromMap(overrideConfigs);
+  auto& writer = createWriter(config);
+  auto& context = getContext();
+  std::array<char, 10> data;
+  std::memset(data.data(), 'a', 10);
+  auto& writerSink = writer.getSink();
+
+  for (size_t i = 0; i < 3; ++i) {
+    writerSink.setMode(WriterSink::Mode::Index);
+    writerSink.addBuffer(*pool_, data.data(), data.size());
+    writerSink.setMode(WriterSink::Mode::Data);
+    writerSink.addBuffer(*pool_, data.data(), data.size());
+    writerSink.setMode(WriterSink::Mode::Footer);
+    writerSink.addBuffer(*pool_, data.data(), data.size());
+    writerSink.setMode(WriterSink::Mode::None);
+    context.stripeRowCount = 123;
+    context.stripeRawSize = 345;
+    addStripeInfo();
+    context.nextStripe();
+  }
+
+  std::string typeStr{"struct<a:int,b:float,c:string>"};
+  HiveTypeParser parser;
+  auto schema = parser.parse(typeStr);
+
+  for (size_t i = 0; i < 2; ++i) {
+    writer.addUserMetadata(
+        folly::to<std::string>(i), folly::to<std::string>(i + 1));
+  }
+  for (size_t i = 0; i < 4; ++i) {
+    getFooter().add_statistics();
+  }
+
+  if (compressionKind_ == CompressionKind::CompressionKind_SNAPPY ||
+      compressionKind_ == CompressionKind::CompressionKind_LZO ||
+      compressionKind_ == CompressionKind::CompressionKind_LZ4 ||
+      compressionKind_ == CompressionKind::CompressionKind_GZIP ||
+      compressionKind_ == CompressionKind::CompressionKind_MAX) {
+    VELOX_ASSERT_THROW(
+        writeFooter(*schema), "Unsupported dwrf compression type");
+    return;
+  }
+
+  writeFooter(*schema);
+  writer.close();
+
+  // deserialize and verify
+  auto reader = createReader();
+
+  auto& ps = reader->getPostScript();
+  ASSERT_EQ(
+      reader->getCompressionKind(),
+      // Verify the compression type is the same as passed-in.
+      // If compression is not passed in (CompressionKind::CompressionKind_MAX),
+      // verify compressionKind the compression type is the same as config.
+      compressionKind_ == CompressionKind::CompressionKind_MAX
+          ? config->get(Config::COMPRESSION)
+          : compressionKind_);
```

Comment:
Is it possible to read back the data and verify? Thanks!

Reply:
So in this test, I replaced the rest of the tests to use `SupportedCompressionTest`, so all the test would write data and read back with
1. no encoding
2. ZLIB encoding
3. ZSTD encoding